### DISCUSSION TITLE
Controller Testing

### DISCUSSION THREAD

**Discussion by djyt** [djyt/cannonball#76] [General]
I've been attempting to improve the GamePad handling in CannonBall. In the past, this has been a weak point, but also one that's difficult for me to rectify. I only have an XBox 360 controller (and a Logitech wheel which is currently in my loft), and don't particularly want to buy loads of others for compatibility testing. And of course, everything works fine for me, but that's not exactly helpful!

There is a beta build here that should be used. Or you can compile your own. Please use the latest config.xml that is included. 

**Edit: Please now use the latest build on the release page.**

**Things to note:**
- I've added gamecontrollerdb.txt support to facilitate the automatic setup of controllers. This file is included in the package. 
- If the padconfig and analog settings in config.xml are set to '-1' (as in the bundled configuration), CannonBall will attempt to assign sensible values to the axis (on controllers with analog sticks and trigger) and buttons.
- You may still need to redefine the buttons to something sensible as all controllers are different.
- You will still need to enable analog controls if you want to use them.
- You can also plug in and detach controllers whilst CannonBall is running now. But note that once a controller is 'configured' it won't try to reconfigure a different one, unless the settings are mapped back to '-1' in the config file. We could add a menu option to setup a gamepad again if this seems useful. But I'm assuming most people just have one gamepad they stick with. 

**Stuff I'd like help on:**
- What model is your controller and does it work out of the box (or at least with minor configuration tweaks)?
- Does the D-Pad work in the menus?
- Do the analog controls work?
- What are your config.xml settings in the controls section if things don't work as expected? 

The more information you can provide the better. I can't promise I'll ever get to a place where every controller ever invented works flawlessly, but at least let's address the low hanging fruit.

**Comment by pjft** [comment]
So, good news. Works fine out of the box for my PS3 sixaxis, starts on digital, if I change it to analog then it uses analog controls.

Well done. Thanks!

**Reply by djyt** [reply]
Cool thanks for the feedback. I've made further changes since the above build, with some help from @DanTheMan827 
Also have basic rumble working too connected to the actual upright wheel motor code. Just gonna build out some nicer menu options for it.

**Comment by baritonomarchetto** [comment]
I have not used the beta version you posted here, but with the one in main page (v0.32-w64) rumble works for me. I have an arduino leonardo emulating an Xbox 360 controller with only one analog axis (wheel) and no dpad (sorry)

**Comment by Kimet1970** [comment]
I'm using a old Microsoft Sidewinder Precision Wheel.
Apparently the FFB is working nice...
But the pedals mapping are inverted!

- Acellerator down - no acceleration
- Acellerator up - maximum acceleration!
- The same bahavour to the brake pedal!

WOuld be possible to fix it?
(Only adding a extra option to invert the pedals counters...)

**Ps.** In the MS Sidewinder Wheel
(Wheel - axis 0, Accelerator - axis 1, brake - axis 3)

**Many thanks to your really great work in Cannonball!**
                                                 Sérgio.

**Reply by djyt** [reply]
Yes, I can easily add an option to invert the axis for you.

**Reply by Kimet1970** [reply]
Many thanks!!!!
(A patch or a future update?)
I think this way, Microsoft Sidewinder will work fine!

**Reply by djyt** [reply]
I'll add it into the next release. Which will probably be within a few weeks.

**Reply by Kimet1970** [reply]
**Many many Thanks djyt!**
Great work
Other 2 possible minor (but cool) implementations:

- A optional gear shifter as MAME
- A cool icon to e.exe

Using Resource Hacker I could to replace the original icon by (another figure - a Cannonball fan made?) with many internal sizes [16x16, 32x32... ...] to exhibition on screen [figure attached]
![Icons](https://user-images.githubusercontent.com/82340648/114421772-08c83f80-9b8c-11eb-9375-a5b06fc63072.jpg)

**Reply by djyt** [reply]
Blimey that icon looks like vomit. But that doesn't mean I shouldn't investigate a higher res one at some point! :-D

**Reply by Kimet1970** [reply]
Really?
For me it's razonable... no very good but "ok".
[No good for "small icoons" options...]
I got that figure in the net...
![image](https://user-images.githubusercontent.com/82340648/114475742-32ef2100-9bcf-11eb-826d-797ce7ea156f.png)

**Reply by djyt** [reply]
> I'm using a old Microsoft Sidewinder Precision Wheel.
> Apparently the FFB is working nice...
> But the pedals mapping are inverted!
> 
> * Acellerator down - no acceleration
> * Acellerator up - maximum acceleration!
> * The same bahavour to the brake pedal!
> 
> WOuld be possible to fix it?
> (Only adding a extra option to invert the pedals counters...)
> 
> **Ps.** In the MS Sidewinder Wheel
> (Wheel - axis 0, Accelerator - axis 1, brake - axis 3)
> 
> **Many thanks to your really great work in Cannonball!**
> Sérgio.

Inverted pedal check-in is here. 96d4af0029b57082004076c3dc31e0368265be2a

I've attached a build below.  Can you see if it fixes your issue?

You will need to setup the config.xml as follows to invert the pedals:
```
<axis>
	<wheel>-1</wheel>
	<accel invert="1">-1</accel>
	<brake invert="1">-1</brake>
</axis>
```
Updated with new build below...
[cannonball_pedal_inversion_v2.zip](https://github.com/djyt/cannonball/files/6360799/cannonball_pedal_inversion_v2.zip)

**Reply by djyt** [reply]
> Really?
> For me it's razonable... no very good but "ok".

Right, rather than simply being a critic, I've dug out an icon someone made me to use. Unfortunately, I think this was sent to me years ago and I cannot for the life of me remember who made it. If you ever read this, please respond and thank you. 

![icon](https://user-images.githubusercontent.com/2414449/114876830-9f983480-9df6-11eb-9543-09d406579a69.PNG)

ae34875d31efbf0ca966d58c0e1c049bee08df02

**Comment by Kimet1970** [comment]
**Many thanks djyt!**
**The sidewinder's pedal is almost working...**
If you press the accelerator until its half course... it's ok!
**But after the half course the accelerations begins to decrease again!**
**Same behavior to the brake... it brakes until the half course... but after it no more breaks the car...**

**The scheme of Sidewinder in Windows 10 is:**
X axis (axis 0 in Cannonball) = wheel (ok!)
Y axis (axis 1 in Cannonball) = accelerator
Z axis (axis 2 in Cannonball) = brake

- The Y axis (accelerator) **goes from a maximum NEGATIVE value (no pressed) until a maximum POSITIVE value (totally pressed)**. **The ZERO value is the half of course**.
- The **Z axis (brake) goes from an maximum POSITIVE value (no pressed) until ZERO (totally pressed)**!

**Ps.** The "-1" option (automatic axis configuration) doesn't work with Sidewinder.

**Are you mapping the module function of the pedal signal output?**
I think... this way... Sidewinder [with "inverse" option activated] will to behavior exactly as descripted above
**Module of signal output decreasing = accelerating & module increasing decelerating...**

Best regards!

![Sidewinder Scheme](https://user-images.githubusercontent.com/82340648/115152524-d44efa80-a047-11eb-90c3-b1e244d608f7.png)

**Reply by djyt** [reply]
Unfortunately, the automatic axis configuration won't work with the SideWinder, that's for sure, as it's not a gamepad. So don't expect that ever to be a feature I'm afraid. 

> Question: Which model / year is your SideWinder from? 

1/ **Steering** - good that's working

2/ **Accelerator** - the code supports analog pedals ranging from between -32768 to 32767.  (GamePads/Controllers usually range from 0 to 32767 and are handled differently if detected, but if the axis remain set at -1, after you've saved the settings from the menu we can assume you need to setup manually)

You state: "The Y axis (accelerator) goes from a maximum NEGATIVE value (no pressed) until a maximum POSITIVE value (totally pressed). The ZERO value is the half of course."

So this should theoretically work, providing, for some reason it isn't being detected as a gamepad. 

This can be inverted as follows:
```<accel invert="1">1</accel>```

> Question: Does it not work correctly with either invert set to 1 or 0 ?

I'm a bit confused as to why the accelerator isn't working correctly with this option. As it would support a pedal ranging from either -32768 to 32767 or, with the invert, 32767 to -32768. Which would seem to cover your description above. 

3/ **Brake** - I'm a bit baffled as to why this axis would be completely different from the Accelerator, but then again, I  have no knowledge of the Sidewinder. This wouldn't be currently supported, I'd need to write a bit of code. But maybe, let's focus on getting the accelerator working first which should be supported with the current code.

**Reply by djyt** [reply]
Can you run the following SDL test application, and provide the console output for:
[joystick_test.zip](https://github.com/djyt/cannonball/files/6334688/joystick_test.zip)

1/ How the device is detected
2/ What happens when accelerator is depressed -> fully pressed (console values)
3/ What happens when accelerator is depressed -> fully pressed (console values)

Thanks

----

Example for say a 360 controller is:

```
INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: Xbox 360 Controller
INFO:        type: Game Controller
INFO:        axes: 6
INFO:       balls: 0
INFO:        hats: 1
INFO:     buttons: 11
INFO: instance id: 0
INFO:        guid: 030000005e0400008e02000000007801
INFO:     VID/PID: 0x045e/0x028e
```

And an example of the output when pressing/depressing the accelerator:

```
INFO: Joystick 0 axis 5 value: 32767
INFO: Joystick 0 axis 5 value: 32510
INFO: Joystick 0 axis 5 value: 29683
INFO: Joystick 0 axis 5 value: 26599
INFO: Joystick 0 axis 5 value: 21459
INFO: Joystick 0 axis 5 value: 16833
INFO: Joystick 0 axis 5 value: 11179
INFO: Joystick 0 axis 5 value: 7324
INFO: Joystick 0 axis 5 value: 3212
INFO: Joystick 0 axis 5 value: -1157
INFO: Joystick 0 axis 5 value: -5526
INFO: Joystick 0 axis 5 value: -9381
INFO: Joystick 0 axis 5 value: -14264
INFO: Joystick 0 axis 5 value: -17091
INFO: Joystick 0 axis 5 value: -20175
INFO: Joystick 0 axis 5 value: -23002
INFO: Joystick 0 axis 5 value: -25058
INFO: Joystick 0 axis 5 value: -27885
INFO: Joystick 0 axis 5 value: -30198
INFO: Joystick 0 axis 5 value: -32254
INFO: Joystick 0 axis 5 value: -32768
```

**Comment by yepyepyep4711** [comment]
Hello!

I'm using an original Playstation 2 DualShock 2 Analog Controller (SCPH-10010), connected to my laptop via a USB adapter. The joystick works in windows both in analog and digital mode. It also works in Retroarch, meaning it works with the lr-cannonball port. But it doesn't work with the current standalone release of cannonball. The mention "gamepad found " does appear when entering the control options, and if the gamepad values are set to -1 in config.xml, they will be updated. But absolutely nothing happens. The controller cannot be used either within the menus or the game.

Let me know if I can provide more information.

EDIT: I've found it (I think). Cannonball says "gamepad found" even if no gamepad is connected. So I guess that imaginary gamepad takes precedence oder the real one.

**Reply by djyt** [reply]
Try the joystick_test.zip file I linked to above and let me know the output. Many Thanks!

**Reply by yepyepyep4711** [reply]
Happy to.

Upon launching the application, an empty black window opens. It disappears when I click on it and the program exits. Is that the expected behaviour?

In any case, here's the output with some comments:

`The lines until the next comment appear as soon as I start the application, with no joystick plugged in`
INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: Monect Hid Device
INFO:        type: Game Controller
INFO:        axes: 0
INFO:       balls: 0
INFO:        hats: 0
INFO:     buttons: 0
INFO: instance id: 1
INFO:        guid: 03000000efbe0000edfe000000000000
INFO:     VID/PID: 0xbeef/0xfeed
INFO: Joystick device 1 added.
`The lines below appear after plugging in the joystick. Since the adapter is a Y cable for two joysticks, the fact that two new joysticks are seen is expected behaviour`
INFO: Joystick device 2 added.
INFO: Joystick device 3 added.

Nothing else appears, regardless of which buttons I press, in either analog or digital mode.

Final remarks: I don't have a clue why my laptop shows a Monect Hid Device for a game controller when there is none. I don't know if that is standard on Windows 7. What I can say is that I also tested an 8bitDo NES30 gamepad connected over USB and achieved the exact same results, both in cannonball and with the joystick test software. That NES30, like the PS2 controller, registers correctly in windows and works flawlessly with other software like Retroarch. Grasping at straws, I'd conjecture that SDL is bypassing windows and grabbing input from the first joystick (device 0), instead of using the default one as specified by the user in the system settings.

**Reply by yepyepyep4711** [reply]
To try and prove my theory, I deactivated all HID devices in the device manager, and lo and behold, both controllers are now first in line and can then be "seen" by both cannonball and the test utility. So the issue with cannonball seems to be that it's taking the first device and not the system preferred device.

Once that hurdle is past, though, I have additional information:
- all buttons on the NES30 work, but the d-pad only registers in the test utility. In cannonball, the d-pad doesn't do anything. Here's the capture from a d-pad test over bluetooth:

INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: Bluetooth Wireless Controller
INFO:        type: Unknown
INFO:        axes: 6
INFO:       balls: 0
INFO:        hats: 1
INFO:     buttons: 16
INFO: instance id: 0
INFO:        guid: 03000000c82d00002028000000000000
INFO:     VID/PID: 0x2dc8/0x2820
INFO: Joystick 0 axis 1 value: -256
INFO: Joystick 0 axis 1 value: 32767
INFO: Joystick 0 axis 2 value: -256
INFO: Joystick 0 axis 2 value: 32767
INFO: Joystick 0 axis 1 value: -256
INFO: Joystick 0 axis 2 value: -256
INFO: Joystick 0 axis 0 value: -256
INFO: Joystick 0 axis 0 value: -32768
INFO: Joystick 0 axis 0 value: -256
INFO: Joystick 0 axis 0 value: 32767
INFO: Joystick 0 axis 0 value: -256

To my amazement, it's different over USB:
INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: 4-Achse 12-Taste Joystick mit Mehrwegeschalter
INFO:        type: Unknown
INFO:        axes: 4
INFO:       balls: 0
INFO:        hats: 1
INFO:     buttons: 12
INFO: instance id: 0
INFO:        guid: 03000000c82d000012ab000000000000
INFO:     VID/PID: 0x2dc8/0xab12
INFO: Joystick 0 axis 1 value: 0
INFO: Joystick 0 axis 1 value: 32767
INFO: Joystick 0 axis 1 value: 0
INFO: Joystick 0 axis 0 value: 0
INFO: Joystick 0 axis 0 value: 32767
INFO: Joystick 0 axis 0 value: 0
INFO: Joystick 0 axis 1 value: -32768
INFO: Joystick 0 axis 1 value: 0
INFO: Joystick 0 axis 0 value: -32768
INFO: Joystick 0 axis 0 value: 0

For comparison, here's the PS2 DualShock over USB:
INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: Twin USB Joystick
INFO:        type: Game Controller
INFO:        axes: 4
INFO:       balls: 0
INFO:        hats: 1
INFO:     buttons: 12
INFO: instance id: 1
INFO:        guid: 03000000100800000100000000000000
INFO:     VID/PID: 0x0810/0x0001
INFO: Joystick device 1 added.
INFO: Joystick 1 axis 1 value: -256
INFO: Joystick 1 axis 1 value: -32768
INFO: Joystick 1 axis 1 value: -256
INFO: Joystick 1 axis 0 value: -256
INFO: Joystick 1 axis 0 value: -32768
INFO: Joystick 1 axis 0 value: -256
INFO: Joystick 1 axis 1 value: 32767
INFO: Joystick 1 axis 1 value: -256
INFO: Joystick 1 axis 0 value: 32767
INFO: Joystick 1 axis 0 value: -256

- lastly, the DualShock works fine in analog mode, but alas, no rumbling :)

Hope this helps.

**Reply by djyt** [reply]
Yes, that is correct that by default it takes the first device. The default can be overridden in the config.xml file as follows:

```
    <!-- Use the N'th joystick on the system. -->
    <pad_id>0</pad_id>
```

I'm not sure if there's a way of getting the system 'default' in SDL. I don't think so unfortunately. I could add a menu option to allow people to select the pad_id, but my assumption is that you'll set this up once and leave it. So it's not really a bug, more just the way things are for the time being. 

Whether or not SDL sees your controller as a 'joystick' or 'gamepad' is dependent on whether it is found in the gamecontrollerDB file and probably other factors that I don't fully understand! 

You could try replacing the one I included with the latest from here and see if it makes any difference:
https://github.com/gabomdq/SDL_GameControllerDB/blob/master/gamecontrollerdb.txt

**Reply by yepyepyep4711** [reply]
Hi again,

right. So I set the pad_id and now the DualShock works without tweaking windows first. The issue with this approach is that people need to know about it in the first place and then find the correct id. My suggestion would be an additional item on the gamepad menu that says "detect gamepad". It would show a message that says "press and hold a button on your gamepad" and poll each existing id (or the first ten if you can't ask the system for existing ids) for a button press, then store that id to the config.

Now to more fun stuff: how can we make rumble work on the DualShock? :)

EDIT: I have successfully installed [these drivers](https://github.com/aitte2/dualshock_driver) and can make the controller rumble using the windows game controller settings for the device.

**Reply by yepyepyep4711** [reply]
I forgot one oddity: using the analog stick for steering works fine, but not for choosing the music. On that screen, it will go full circle and keep rotating through the three music tracks.

**Comment by RobWW** [comment]
My thanks as well djyt!

I am testing with a new Logitech G920 wheel.
I am using the latest version from this thread:  'cannonball_pedal_inversion'
Out of the box, the D-Pad and buttons are detected, and I can use the D-Pad to navigate the menus.

To get the analogue controls to work, I had to configure the axis manually and inversion was required for the accelerator and brake:
    `<wheel>0</wheel>`
    `<accel invert="1">1</accel>`
    `<brake invert="1">2</brake>`

However, I noticed a problem with the accelerator:  while working for most of the range, when pressed to maximum it behaves as if not pressed at all, that is the car will stop.
'Joystick Test' reports values from the accelerator:
Unpressed:  INFO: Joystick 0 axis 1 value: 32767
Fully down:  INFO: Joystick 0 axis 1 value: -32768  
The accelerator behaves as expected for most of the range.  It is only when it is close to or at the maximum value of -32768 that the problem occurs.

One other observation: in terms of game play, the brake is not very sensitive and the pedal has to be depressed about half way before the brakes are activated in the game.  This can be worked around by adjusting the Windows settings, but it would be nice if this threshold was tunable.
'Joystick Test' reports values from the brake:
Unpressed:  INFO: Joystick 0 axis 2 value: 32767
Fully down:  INFO: Joystick 0 axis 2 value: -32768

If it helps, here is the output from Joystick Test
INFO: Joystick device 0 added.
INFO: Joystick
INFO:        name: Logitech G920 Driving Force Racing Wheel USB
INFO:        type: Wheel
INFO:        axes: 4
INFO:       balls: 0
INFO:        hats: 1
INFO:     buttons: 18
INFO: instance id: 0
INFO:        guid: 030000006d04000062c2000000000000
INFO:     VID/PID: 0x046d/0xc262

I will be happy to do some more testing.
Best regards!

**Reply by djyt** [reply]
Right.

1/ I found a bug with the way in which I was handling pedal inversion. Please try this out. I have simulated this as much as possible without actually having the right controller. It should work for you.

2/ I also realized the controllerdb was missing from the released I previously provided. It's included in this one. This may have impacted detecting controllers and rumble. 

[cannonball_pedal_inversion_v2.zip](https://github.com/djyt/cannonball/files/6360794/cannonball_pedal_inversion_v2.zip)

**Reply by RobWW** [reply]
Hi djyt,thanks for the new build.

1/  The accelerator pedal is now working correctly.
2/  For controller detection, with the default configuration the D-pad is detected and the program reports 'Gamepad found'.  I had to use the 'Configure Gamepad' to set up the remaining keys, which I assume is expected.  This set up also let me correctly define the the accelerator and brake.  However, the steering wheel is not detected.  I had to manually set the steering to 0 and, of course, invert the accelerator and brake.

 So now the game is completely playable with the G920.  Thanks!

I did noticed, however, that neither the rumble effect nor the haptic force feedback seemed to work, in that I could not feel a difference if they were enabled or disabled.  I tried the haptic feedback with both positive and negative values. 

I will be happy to continue testing!

**Reply by djyt** [reply]
That's good news Rob. Yes, only supported controllers get auto-configured. Things like wheels and custom pedals will always be a manual effort unfortunately.

My wheel is currently in the loft, so I haven't messed around with this area of code in a while. I am not too sure why the force feedback wouldn't be working unfortunately. I do want to port that code to SDL 2 at some point (it's currently using some ancient DirectX code). This should improve compatibility and mean it's fully cross platform.

**Comment by Kimet1970** [comment]
![WhatsApp Image 2021-04-24 at 05 32 26](https://user-images.githubusercontent.com/82340648/115952974-654a2980-a4bf-11eb-9afb-210747012f77.jpeg)
**Hi djyt!**
After your "pedal inversion bug" fix...
My Microsoft Sidewinder Force Feedback Wheel is working fine(?)!

Steering and FFB apparently working fine (ok!)
Pedals working fine(?) but apparently in a "digital mode". (ok?)

The pedals are working and they are perfectly playable!
But I can't e.g. to press the accelarator until a little fraction of its total course and go continually in low speed...
The car accelerates continually (until the maximum speed...
...and [I think] the acceleration rate isn't related to pedal position...
The same behavour to the brake... I can brake, but I can't to do smoothly breakings [I only achieve it doing a brief time brake pressure]...
**Are the pedal digitals in the original game?**
**[Or it is it related to the "pedal speed" parameter in the config file? I can't understood it very well... Is it a response speed? A lag in the response?]**
**When I plug my Sidewinder Wheel I can see "GAMEPAD FOUND"!
Is Canonnball seeing my Sidewinder Wheel (or my Sidewinder pedals) as a gamepad?**

Many thanks to you!! 
Best regards!

**Reply by djyt** [reply]
The pedals are analog in the original game. If you set 
```<analog enabled="1">```
then all pedals should be analog if your hardware supports it. 

The digital pedal speed is for digital controls to simulate analog. (i.e. you press a digital button and it slowly emulates fully pressing the pedal). However, this is only used when using digital controls. 

You need to run the joystick test app, as per my previous post, and provide the results in here. There are way too many variables to guess what the problem might be. 
https://github.com/djyt/cannonball/discussions/76#discussioncomment-629199

**Comment by yepyepyep4711** [comment]
Regarding rumble and FFB, maybe the work on the [FFB Arcade Plugin](https://github.com/Boomslangnz/FFBArcadePlugin) could be interesting. It uses SDL2 too if I'm not mistaken.

**Reply by djyt** [reply]
That's coded in DirectX - which is what I want to avoid as it locks it to Windows machines. The only reason I coded the FFB support using DirectX is that back in ~2013 when I coded it, SDL didn't support haptic/FFB devices.

**Reply by yepyepyep4711** [reply]
damn. Right. It seems the author is only currently moving to sdl2. See the [comments to the release before last](https://github.com/Boomslangnz/FFBArcadePlugin/releases/tag/2.0.0.5). If the author is right about the WIP status of rumble in SDL2, that could explain inconsistency in the results we here get.

**Comment by Kimet1970** [comment]
**Hi djyt!**
Actually, in Windows 10 there are **2 possible "working modes"** to Microsoft Sidewinder FFB Wheel!
**If you calibrate the wheel** (by "Control Panel/... ..."), it will work with accelerator (axis 1) varying from 0 to -32768 (when totally pressed) and the brake varying from +32767 to -32768 (when totally pressed).
*_(Values obtained using your "joystick_test")._

But if you didn't calibate it or **clicking in "Reset to default"** using "Control Panel/hardware and sound/devices and printers/Sidewinder Force Feedback Wheel (USB)/properties...

The accelerator will work now varying from +32767 to -32768 (when tottaly pressed)! 

Until to calibrate it again!

For the "Reset to default" option, the values from joystick_test were: 

![Forum Figuras](https://user-images.githubusercontent.com/82340648/116200667-fe22b400-a70e-11eb-91e4-4cbaaf655e55.png)

**Reply by djyt** [reply]
Cool, so just to confirm - that should work then if you configure the axis correctly in the config.xml, based on what you've posted above?

**Reply by Kimet1970** [reply]
**Yes! It's working fine!

**Only a question...**
If I enable the option "Haptic Support" I loses some FFB effects as vibrations in the wheel when driving off-road... (Is it normal?)
And I can't to see any quality increase (or even any modification) in the general FFB behaviour...

Many thanks!

**Reply by djyt** [reply]
Yes that's normal. 

Haptic = Force Feedback. This uses the deluxe sit-down motor code to provide resistance to the wheel. The original sit-down cabinet did not have a steering wheel shaker when driving off-road. Unfortunately, this is possible an unfortunate use of terminology due to also having rumble. 

Rumble = Very basic vibration. This uses the upright motor code which shakes the steering wheel when driving off-road and crashing on supported gamepads.

**Reply by Kimet1970** [reply]
Ok.
But if Haptic uses the deluxe sit-down motor code (with steering wheel shake implemented)...
I would hope that:
Haptic <1> = with wheel shaker when driving
Haptic <0> = without wheel shaker when driving

But, actually, just the opposite occurs...
**Am I missing something?**

**Reply by djyt** [reply]
If your wheel supports force feedback, haptic should enable force feedback. It's possible it's not compatible if it doesn't work. I've only tested on the wheel I own.

**Reply by Kimet1970** [reply]
Ok.
In fact, I suspect I only getting a default directed to center constant force and some (when off-road) vibrations (from rumble effect). And no vibrations or forces in the collisions.
I can't to rescale the returning force it using changing the parameters max and min force or even to perceiving a dynamical behaviour on it...
I think I getting something as a old times FFB...
_**(*In Microsoft FFB Wheel, only plugging it in the power outlet and USB we already get a constant (non variable) to center force... I don't know the working way of other wheels...)**_
After all... it isn't bad.
If I don't missing anything... One (sitdown deluxe?) of the originals Outrun hardware (a game from 80's) had a dynamical returning to center force? (!!) 
It's very surprising to me... I would expect only a primitive constant force or something in this line... exactly as I getting...

**Reply by djyt** [reply]
Nope. All the original OutRun hardware simply centered the wheel using a pair of strong springs. 
There was no force feedback or anything particularly high-tech in the original machines.

**Reply by yepyepyep4711** [reply]
> Rumble = Very basic vibration. This uses the upright motor code which shakes the steering wheel when driving off-road and crashing on supported gamepads.

Hi, I wanted to ask: what does "supported gamepads" mean? In my mind, anything that rumbles in windows should be supported, but maybe that's different here. Do you need to actively code for each hardware, or is there some abstraction layer you can tap into?

**Reply by djyt** [reply]
Supported gamepads means detected and compatible with SDL in this particular case, which is the abstraction layer.

**Reply by yepyepyep4711** [reply]
right. I'm still trying to pinpoint where something goes missing.
- My controller works to steer in the game
- It can rumble using the dialogue under Devices -> Game controller options in windows

So why doesn't it also rumble in the game? I've tried looking for some software to test rumble using sdl2 but came up empty so far, except for [this](https://gist.github.com/egore/4721bc21047d165c8bc47dfa7b848775), and I don't know if it's any help.

**Reply by djyt** [reply]
It's possible it isn't fully compatible with SDL as a device that supports rumble, despite working in Windows. Exactly why - I unfortunately don't know. There are thousands of gamepads out there, and I'm by no means an expert on all of their idiosyncrasies.  

The test code you found would be a good test-case if you could compile and run it.

**Reply by Kimet1970** [reply]
**Testing the old version 0.3 (from 2014) posted in the "Issues" section...**
Contrary the last recent versions...
**The Force Feedback works fine in my older Microsoft Forcefeedback!!**
Selecting haptic=1 I can to select the maximum and minimum force... and to obtain the "rumbles" when offroad...
Apparently all working fine in 0.3 version from 2014!!
But it (FFB) doesn't work in recent version...
Only the constant standart force to center in the wheel... :( 
A possible bug in the last versions code?
Or maybe a incompatibility between 64 bits software and my old Sidewinder?

**Reply by djyt** [reply]
Thanks for the feedback. I'll use the Issue thread to discuss further as it's confirmed an 'issue'. Cheers for testing for me.

**Reply by DanTheMan827** [reply]
A workaround might be to put cannonball into steam, go into big picture, and turn on the options related to gamepad emulation.

That will make just about anything show up as a standard XInput controller and should be recognized by cannonball as such.

**Reply by yepyepyep4711** [reply]
> A workaround might be to put cannonball into steam, go into big picture, and turn on the options related to gamepad emulation.
> 
> That will make just about anything show up as a standard XInput controller and should be recognized by cannonball as such.

Not a fan of needing an account for everything I do, so personally I'd rather this not be contemplated.

**Comment by patoperales** [comment]
Good morning everyone. 

I'm working on a project that involves using Cannonball on my Raspberry Pi 4 running RetroPie. 
My idea is to use SmartyPi's `enable` function to capture console outputs for motor motion and lights and send them to an Arduino. 
I have a Cisco Heat arcade cabinet whose steering wheel has a vibration motor and driver board. My plan is to send the Arduino's output to the driver board to simulate the behavior of the Uipright version of OutRun...so far so good.
My concern is about the controls...
For the analog controls, I'm using an Ultimarc board called APAC. Does anyone have experience with this or know if Cannonball correctly recognizes the APAC as an analog controller for the steering wheel and pedals?

THANK YOUUU
